gateway-service-impl/src/main/java/org/hypertrace/gateway/service/baseline/BaselineServiceImpl.java

```diff
@@ -0,0 +1,383 @@
+        baselineEntityMap.put(baselineEntity.getId(), baselineEntityBuilder.build());
+      }
+    }
+    return baselineEntityMap;
+  }
+
+  private List<TimeAggregation> getTimeAggregationsForTimeSeriesExpr(
+      BaselineEntitiesRequest originalRequest) {
+    List<BaselineTimeAggregation> timeSeriesList =
+        originalRequest.getBaselineMetricSeriesRequestList();
+    List<TimeAggregation> timeAggregations = new ArrayList<>();
+    for (BaselineTimeAggregation timeAggregation : timeSeriesList) {
+      Expression aggregation =
+          Expression.newBuilder().setFunction(timeAggregation.getAggregation()).build();
+      TimeAggregation agg =
+          TimeAggregation.newBuilder()
+              .setAggregation(aggregation)
+              .setPeriod(timeAggregation.getPeriod())
+              .build();
+      timeAggregations.add(agg);
+    }
+    return timeAggregations;
+  }
+
+  private List<TimeAggregation> getTimeAggregationsForAggregateExpr(
+      BaselineEntitiesRequest originalRequest) {
+    List<FunctionExpression> aggregateList = originalRequest.getBaselineAggregateRequestList();
+    List<TimeAggregation> timeAggregationList = new ArrayList<>();
+    for (FunctionExpression function : aggregateList) {
+      TimeAggregation timeAggregation =
+          getAggregationFunction(
+              function, originalRequest.getStartTimeMillis(), originalRequest.getEndTimeMillis());
+      timeAggregationList.add(timeAggregation);
+    }
+    return timeAggregationList;
+  }
+
+  private TimeAggregation getAggregationFunction(
+      FunctionExpression function, long startTimeMillis, long endTimeMillis) {
+    Period period = getPeriod(startTimeMillis, endTimeMillis);
+    TimeAggregation timeAggregation =
+        TimeAggregation.newBuilder()
+            .setPeriod(period)
+            .setAggregation(Expression.newBuilder().setFunction(function).build())
+            .build();
+    return timeAggregation;
+  }
+
+  private Period getPeriod(long startTimeInMillis, long endTimeMillis) {
+    Period period;
+    if ((endTimeMillis - startTimeInMillis) < DAY_IN_MILLIS) {
+      long timePeriod = (endTimeMillis - startTimeInMillis) / (1000 * 60);
+      period = Period.newBuilder().setUnit("MINUTES").setValue((int) timePeriod).build();
+    } else {
+      long timePeriod = (endTimeMillis - startTimeInMillis) / (1000 * 60 * 60 * 24);
+      period = Period.newBuilder().setUnit("HOURS").setValue((int) timePeriod).build();
+    }
+    return period;
+  }
+
+  private Period getTimeSeriesPeriod(List<BaselineTimeAggregation> baselineTimeSeriesList) {
+    return baselineTimeSeriesList.get(0).getPeriod();
+  }
+
+  private long getPeriodInSecs(Period period) {
+    ChronoUnit unit = ChronoUnit.valueOf(period.getUnit());
+    long periodSecs = Duration.of(period.getValue(), unit).getSeconds();
+    return periodSecs;
+  }
+
+  /**
+   * If user selected time range is less than 24 hrs then choose last 24hrs of data for baseline if
```

Comment:
would recommend max of (24h, 2xTR) - to keep things continuous (i.e. don't jump from 24h of data to 48h exactly at the 1d mark.

Reply:
Not sure what you mean - if I ask for, say, 23h of data - the baseline will be calculated using 24h of data (and with my other comment - it actually means it will be calculated with only one hour of historical data, since we're calculate the baseline time range incorrectly based on the time range end). On the other hand, asking for 24h of data, the baseline is calculated using 48h (but again, actually only t_0-24h)

Reply:
I agree your point, but in UI you cannot select 23 hours, either you can select 12 hours or 24 hours. But anyway this is small change i will do it

Reply:
Those are the shortcut relative time ranges - the UI also supports any custom time range

Reply:
Fixed this